core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```diff
@@ -0,0 +1,121 @@
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.util;
+
+import org.apache.lucene.util.IOUtils;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.xcontent.ToXContent;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.gateway.MetaDataStateFormat;
+import org.elasticsearch.gateway.MetaStateService;
+import org.elasticsearch.index.Index;
+import org.elasticsearch.index.IndexSettings;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardCopyOption;
+
+/**
+ * Renames index folders from {index.name} to {index.uuid}
+ */
+public class IndexFolderUpgrader {
+    private final NodeEnvironment nodeEnv;
+    private final static ESLogger logger = Loggers.getLogger(IndexFolderUpgrader.class);
+
+
+    /**
+     * Creates a new upgrader instance
+     * @param nodeEnv the node env to operate on
+     */
+    IndexFolderUpgrader(NodeEnvironment nodeEnv) {
+        this.nodeEnv = nodeEnv;
+    }
+
+    /**
+     * Renames all data paths including custom data paths
+     */
+    void upgrade(final IndexSettings indexSettings) throws IOException {
+        final Index index = indexSettings.getIndex();
+        for (NodeEnvironment.NodePath nodePath : nodeEnv.nodePaths()) {
+            upgrade(index, nodePath.indicesPath.resolve(index.getName()));
+        }
+        if (indexSettings.hasCustomDataPath()) {
+            upgrade(index, nodeEnv.resolveCustomLocation(indexSettings).resolve(index.getName()));
+        }
+    }
+
+    /**
+     * Renames the index found in <code>sourceDir</code>
+     */
+    void upgrade(final Index index, final Path sourceDir) throws IOException {
+        final Path targetDir = sourceDir.resolveSibling(index.getUUID());
+        if (Files.exists(sourceDir)) {
+            logger.info("{} moving from [{}] to [{}]", index, sourceDir, targetDir);
```

Comment:
can we add something that will explain why we're doing this of the unordained user? something ala upgrading indexing folder to new naming conventions